diff --git a/Configurations.md b/Configurations.md
index fce37eec..f9b623bc 100644
--- a/Configurations.md
+++ b/Configurations.md
@@ -2345,7 +2345,7 @@ Controls the strategy for how consecutive imports are grouped together.
 Controls the strategy for grouping sets of consecutive imports. Imports may contain newlines between imports and still be grouped together as a single set, but other statements between imports will result in different grouping sets.
 
 - **Default value**: `Preserve`
-- **Possible values**: `Preserve`, `StdExternalCrate`, `One`
+- **Possible values**: `Preserve`, `StdExternalCrate`, `One`, `ByDistance`, `ByDistanceDescending`
 - **Stable**: No (tracking issue: [#5083](https://github.com/rust-lang/rustfmt/issues/5083))
 
 Each set of imports (one or more `use` statements, optionally separated by newlines) will be formatted independently. Other statements such as `mod ...` or `extern crate ...` will cause imports to not be grouped together.
@@ -2410,6 +2410,82 @@ use std::sync::Arc;
 use uuid::Uuid;
 ```
 
+#### `ByDistance`
+
+Discard existing import groups, and create four groups for:
+1. `self`,
+2. `super`,
+3. `crate`,
+4. external crates.
+
+This option variant is not compatible with `import_granularity` as the individual groups are normalized after regrouping.
+
+```rust
+mod dolor;
+mod ipsum;
+
+use self::{dolor::foo, ipsum::bar};
+
+use super::{
+    schema::{Context, Payload},
+    update::convert_publish_payload,
+};
+
+use crate::models::Event;
+
+use {
+    alloc::alloc::Layout,
+    broker::database::PooledConnection,
+    chrono::Utc,
+    core::f32,
+    juniper::{FieldError, FieldResult},
+    std::sync::Arc,
+    uuid::Uuid,
+};
+```
+
+#### `ByDistanceDescending`
+
+Discard existing import groups, and create four groups for:
+1. external crates,
+2. `crate`,
+3. `super`,
+4. `self`.
+
+This option variant is not compatible with `import_granularity` as the individual groups are normalized after regrouping.
+
+```rust
+mod dolor;
+mod ipsum;
+
+use {
+    alloc::alloc::Layout,
+    broker::database::PooledConnection,
+    chrono::Utc,
+    core::f32,
+    juniper::{FieldError, FieldResult},
+    std::sync::Arc,
+    uuid::Uuid,
+};
+
+use crate::models::Event;
+
+use super::{
+    schema::{Context, Payload},
+    update::convert_publish_payload,
+};
+
+use self::{dolor::foo, ipsum::bar};
+```
+
+## `regroup_modules`
+
+Discard all newlines between consecutive module declarations and group them into a single consecutive group.
+
+- **Default value**: `false`
+- **Possible values**: `true`, `false`
+- **Stable**: No
+
 ## `reorder_modules`
 
 Reorder `mod` declarations alphabetically in group.
diff --git a/src/config/mod.rs b/src/config/mod.rs
index b03674b6..82849907 100644
--- a/src/config/mod.rs
+++ b/src/config/mod.rs
@@ -102,6 +102,8 @@ create_config! {
     // Ordering
     reorder_imports: ReorderImports, true, "Reorder import and extern crate statements \
         alphabetically";
+    regroup_modules: RegroupModules, false, "Discard newlines and regroup all consecutive modules \
+        together";
     reorder_modules: ReorderModules, true, "Reorder module statements alphabetically in group";
     reorder_impl_items: ReorderImplItems, false, "Reorder impl items";
 
@@ -790,6 +792,7 @@ imports_layout = "Mixed"
 imports_granularity = "Preserve"
 group_imports = "Preserve"
 reorder_imports = true
+regroup_modules = false
 reorder_modules = true
 reorder_impl_items = false
 type_punctuation_density = "Wide"
@@ -882,6 +885,7 @@ imports_layout = "Mixed"
 imports_granularity = "Preserve"
 group_imports = "Preserve"
 reorder_imports = true
+regroup_modules = false
 reorder_modules = true
 reorder_impl_items = false
 type_punctuation_density = "Wide"
diff --git a/src/config/options.rs b/src/config/options.rs
index 00f9c3f7..d23d2b38 100644
--- a/src/config/options.rs
+++ b/src/config/options.rs
@@ -118,6 +118,18 @@ pub enum GroupImportsTactic {
     StdExternalCrate,
     /// Discard existing groups, and create a single group for everything
     One,
+    /// Discard existing groups, and create new groups for
+    /// 1. submodules / `self` import
+    /// 2. `super` imports
+    /// 3. `crate` imports
+    /// 4. other imports
+    ByDistance,
+    /// Discard existing groups, and create new groups for
+    /// 1. other imports
+    /// 2. `crate` imports
+    /// 3. `super` imports
+    /// 4. submodules / `self` import
+    ByDistanceDescending,
 }
 
 #[config_type]
@@ -646,6 +658,7 @@ config_option_with_style_edition_default!(
 
     // Ordering
     ReorderImports, bool, _ => true;
+    RegroupModules, bool, _ => false;
     ReorderModules, bool, _ => true;
     ReorderImplItems, bool, _ => false;
 
diff --git a/src/reorder.rs b/src/reorder.rs
index e9d0fc05..ffa676eb 100644
--- a/src/reorder.rs
+++ b/src/reorder.rs
@@ -7,13 +7,14 @@
 // FIXME(#2455): Reorder trait items.
 
 use std::cmp::Ordering;
+use std::collections::HashSet;
 
 use rustc_ast::{ast, attr};
 use rustc_span::{Span, symbol::sym};
 
 use crate::StyleEdition;
-use crate::config::{Config, GroupImportsTactic};
-use crate::imports::{UseSegmentKind, UseTree, normalize_use_trees_with_granularity};
+use crate::config::{Config, GroupImportsTactic, ImportGranularity};
+use crate::imports::{UseSegment, UseSegmentKind, UseTree, normalize_use_trees_with_granularity};
 use crate::items::{is_mod_decl, rewrite_extern_crate, rewrite_mod};
 use crate::lists::{ListFormatting, ListItem, itemize_list, write_list};
 use crate::rewrite::{RewriteContext, RewriteError, RewriteResult};
@@ -134,6 +135,13 @@ fn rewrite_reorderable_or_regroupable_items(
                     vec![normalized_items]
                 }
                 GroupImportsTactic::StdExternalCrate => group_imports(normalized_items),
+                GroupImportsTactic::ByDistance | GroupImportsTactic::ByDistanceDescending => {
+                    group_imports_by_distance(
+                        normalized_items,
+                        context.visited_mod_idents,
+                        context.config.group_imports(),
+                    )
+                }
             };
 
             if context.config.reorder_imports() {
@@ -222,6 +230,65 @@ fn group_imports(uts: Vec<UseTree>) -> Vec<Vec<UseTree>> {
     vec![std_imports, external_imports, local_imports]
 }
 
+/// Divides imports into four groups based on the distance to the current
+/// module. Normalizes and sorts each subgroup into a single `use ...` item according to the
+/// `group_imports` tactic.
+fn group_imports_by_distance(
+    uts: Vec<UseTree>,
+    visited_mod_idents: &HashSet<String>,
+    group_imports_tactic: GroupImportsTactic,
+) -> Vec<Vec<UseTree>> {
+    let mut local_use = Vec::new();
+    let mut super_use = Vec::new();
+    let mut crate_use = Vec::new();
+    let mut external_use = Vec::new();
+
+    for mut ut in uts {
+        if ut.path.is_empty() {
+            external_use.push(ut);
+            continue;
+        }
+
+        match &ut.path[0].kind {
+            UseSegmentKind::Slf(_) => local_use.push(ut),
+            UseSegmentKind::Ident(id, _) => {
+                if visited_mod_idents.contains(id.as_str()) {
+                    // Ugly because not efficient: Avoid cloning the `UseTree` by inserting the
+                    // `self` segment in place, but change the rewrite algorithm of the `UseTree`.
+                    ut.path.insert(
+                        0,
+                        UseSegment {
+                            kind: UseSegmentKind::Slf(None),
+                            style_edition: ut.path[0].style_edition,
+                        },
+                    );
+                    local_use.push(ut)
+                } else {
+                    external_use.push(ut)
+                }
+            }
+            UseSegmentKind::Super(_) => super_use.push(ut),
+            UseSegmentKind::Crate(_) => crate_use.push(ut),
+            UseSegmentKind::Glob | UseSegmentKind::List(_) => external_use.push(ut),
+        }
+    }
+
+    let local_use = normalize_use_trees_with_granularity(local_use, ImportGranularity::One);
+    let super_use = normalize_use_trees_with_granularity(super_use, ImportGranularity::One);
+    let crate_use = normalize_use_trees_with_granularity(crate_use, ImportGranularity::One);
+    let external_use = normalize_use_trees_with_granularity(external_use, ImportGranularity::One);
+
+    match group_imports_tactic {
+        GroupImportsTactic::ByDistance => {
+            vec![local_use, super_use, crate_use, external_use]
+        }
+        GroupImportsTactic::ByDistanceDescending => {
+            vec![external_use, crate_use, super_use, local_use]
+        }
+        _ => unreachable!(),
+    }
+}
+
 /// A simplified version of `ast::ItemKind`.
 #[derive(Debug, PartialEq, Eq, Copy, Clone)]
 enum ReorderableItemKind {
@@ -250,6 +317,7 @@ impl ReorderableItemKind {
         ReorderableItemKind::from(item) == self
     }
 
+    /// Whether items of this kind should be reordered.
     fn is_reorderable(self, config: &Config) -> bool {
         match self {
             ReorderableItemKind::ExternCrate => config.reorder_imports(),
@@ -259,18 +327,20 @@ impl ReorderableItemKind {
         }
     }
 
+    /// Whether items of this kind should be regrouped.
     fn is_regroupable(self, config: &Config) -> bool {
         match self {
-            ReorderableItemKind::ExternCrate
-            | ReorderableItemKind::Mod
-            | ReorderableItemKind::Other => false,
+            ReorderableItemKind::ExternCrate | ReorderableItemKind::Other => false,
+            ReorderableItemKind::Mod => config.regroup_modules(),
             ReorderableItemKind::Use => config.group_imports() != GroupImportsTactic::Preserve,
         }
     }
 
+    /// Whether items of this kind should be considered as a group when separated by newlines.
     fn in_group(self, config: &Config) -> bool {
         match self {
-            ReorderableItemKind::ExternCrate | ReorderableItemKind::Mod => true,
+            ReorderableItemKind::ExternCrate => true,
+            ReorderableItemKind::Mod => !config.regroup_modules(),
             ReorderableItemKind::Use => config.group_imports() == GroupImportsTactic::Preserve,
             ReorderableItemKind::Other => false,
         }
@@ -329,6 +399,18 @@ impl<'b, 'a: 'b> FmtVisitor<'a> {
     /// Visits and format the given items. Items are reordered If they are
     /// consecutive and reorderable.
     pub(crate) fn visit_items_with_reordering(&mut self, mut items: &[&ast::Item]) {
+        if matches!(
+            self.config.group_imports(),
+            GroupImportsTactic::ByDistance | GroupImportsTactic::ByDistanceDescending
+        ) {
+            self.visited_mod_idents.clear();
+            self.visited_mod_idents
+                .extend(items.iter().filter_map(|ppi| match (**ppi).kind {
+                    ast::ItemKind::Mod(_, ident, _) => Some(ident.name.to_string()),
+                    _ => None,
+                }));
+            println!("Visited mod idents: {:#?}", self.visited_mod_idents);
+        }
         while !items.is_empty() {
             // If the next item is a `use`, `extern crate` or `mod`, then extract it and any
             // subsequent items that have the same item kind to be reordered within
diff --git a/src/rewrite.rs b/src/rewrite.rs
index 7ca1ca52..22ead8a8 100644
--- a/src/rewrite.rs
+++ b/src/rewrite.rs
@@ -1,6 +1,7 @@
 // A generic trait to abstract the rewriting of an element (of the AST).
 
 use std::cell::{Cell, RefCell};
+use std::collections::HashSet;
 use std::rc::Rc;
 
 use rustc_ast::ptr;
@@ -123,6 +124,7 @@ pub(crate) struct RewriteContext<'a> {
     pub(crate) report: FormatReport,
     pub(crate) skip_context: SkipContext,
     pub(crate) skipped_range: Rc<RefCell<Vec<(usize, usize)>>>,
+    pub(crate) visited_mod_idents: &'a HashSet<String>,
 }
 
 pub(crate) struct InsideMacroGuard {
diff --git a/src/visitor.rs b/src/visitor.rs
index d7b1178a..053caef7 100644
--- a/src/visitor.rs
+++ b/src/visitor.rs
@@ -1,4 +1,5 @@
 use std::cell::{Cell, RefCell};
+use std::collections::HashSet;
 use std::rc::Rc;
 use std::sync::Arc;
 
@@ -87,6 +88,7 @@ pub(crate) struct FmtVisitor<'a> {
     pub(crate) report: FormatReport,
     pub(crate) skip_context: SkipContext,
     pub(crate) is_macro_def: bool,
+    pub(crate) visited_mod_idents: HashSet<String>,
 }
 
 impl<'a> Drop for FmtVisitor<'a> {
@@ -807,6 +809,7 @@ impl<'b, 'a: 'b> FmtVisitor<'a> {
             skipped_range: Rc::new(RefCell::new(vec![])),
             is_macro_def: false,
             macro_rewrite_failure: false,
+            visited_mod_idents: HashSet::new(),
             report,
             skip_context,
         }
@@ -1026,6 +1029,7 @@ impl<'b, 'a: 'b> FmtVisitor<'a> {
             report: self.report.clone(),
             skip_context: self.skip_context.clone(),
             skipped_range: self.skipped_range.clone(),
+            visited_mod_idents: &self.visited_mod_idents,
         }
     }
 }
diff --git a/tests/source/configs/group_imports/ByDistance.rs b/tests/source/configs/group_imports/ByDistance.rs
new file mode 100644
index 00000000..88fa075f
--- /dev/null
+++ b/tests/source/configs/group_imports/ByDistance.rs
@@ -0,0 +1,21 @@
+// rustfmt-group_imports: ByDistance
+
+mod ipsum;
+mod dolor;
+
+use chrono::Utc;
+use super::update::convert_publish_payload;
+
+use juniper::{FieldError, FieldResult};
+use uuid::Uuid;
+use ipsum::bar;
+use alloc::alloc::Layout;
+
+use std::sync::Arc;
+
+use broker::database::PooledConnection;
+use self::dolor::foo;
+
+use super::schema::{Context, Payload};
+use core::f32;
+use crate::models::Event;
diff --git a/tests/source/configs/group_imports/ByDistanceDescending.rs b/tests/source/configs/group_imports/ByDistanceDescending.rs
new file mode 100644
index 00000000..cff794c4
--- /dev/null
+++ b/tests/source/configs/group_imports/ByDistanceDescending.rs
@@ -0,0 +1,21 @@
+// rustfmt-group_imports: ByDistanceDescending
+
+mod ipsum;
+mod dolor;
+
+use chrono::Utc;
+use super::update::convert_publish_payload;
+
+use juniper::{FieldError, FieldResult};
+use uuid::Uuid;
+use ipsum::bar;
+use alloc::alloc::Layout;
+
+use std::sync::Arc;
+
+use broker::database::PooledConnection;
+use self::dolor::foo;
+
+use super::schema::{Context, Payload};
+use core::f32;
+use crate::models::Event;
diff --git a/tests/source/configs/group_imports/dolor/mod.rs b/tests/source/configs/group_imports/dolor/mod.rs
new file mode 100644
index 00000000..e69de29b
diff --git a/tests/source/configs/group_imports/ipsum/mod.rs b/tests/source/configs/group_imports/ipsum/mod.rs
new file mode 100644
index 00000000..e69de29b
diff --git a/tests/source/configs/regroup_modules/dolor/mod.rs b/tests/source/configs/regroup_modules/dolor/mod.rs
new file mode 100644
index 00000000..8b137891
--- /dev/null
+++ b/tests/source/configs/regroup_modules/dolor/mod.rs
@@ -0,0 +1 @@
+
diff --git a/tests/source/configs/regroup_modules/false.rs b/tests/source/configs/regroup_modules/false.rs
new file mode 100644
index 00000000..b323e079
--- /dev/null
+++ b/tests/source/configs/regroup_modules/false.rs
@@ -0,0 +1,8 @@
+// rustfmt-regroup_modules: false
+// Regroup modules
+
+mod lorem;
+mod dolor;
+
+pub mod ipsum;
+mod sit;
diff --git a/tests/source/configs/regroup_modules/ipsum/mod.rs b/tests/source/configs/regroup_modules/ipsum/mod.rs
new file mode 100644
index 00000000..8b137891
--- /dev/null
+++ b/tests/source/configs/regroup_modules/ipsum/mod.rs
@@ -0,0 +1 @@
+
diff --git a/tests/source/configs/regroup_modules/lorem/mod.rs b/tests/source/configs/regroup_modules/lorem/mod.rs
new file mode 100644
index 00000000..8b137891
--- /dev/null
+++ b/tests/source/configs/regroup_modules/lorem/mod.rs
@@ -0,0 +1 @@
+
diff --git a/tests/source/configs/regroup_modules/sit/mod.rs b/tests/source/configs/regroup_modules/sit/mod.rs
new file mode 100644
index 00000000..8b137891
--- /dev/null
+++ b/tests/source/configs/regroup_modules/sit/mod.rs
@@ -0,0 +1 @@
+
diff --git a/tests/source/configs/regroup_modules/true.rs b/tests/source/configs/regroup_modules/true.rs
new file mode 100644
index 00000000..f7630286
--- /dev/null
+++ b/tests/source/configs/regroup_modules/true.rs
@@ -0,0 +1,8 @@
+// rustfmt-regroup_modules: true
+// Regroup modules
+
+mod lorem;
+mod dolor;
+
+pub mod ipsum;
+mod sit;
diff --git a/tests/target/configs/group_imports/ByDistance.rs b/tests/target/configs/group_imports/ByDistance.rs
new file mode 100644
index 00000000..7418791a
--- /dev/null
+++ b/tests/target/configs/group_imports/ByDistance.rs
@@ -0,0 +1,23 @@
+// rustfmt-group_imports: ByDistance
+
+mod dolor;
+mod ipsum;
+
+use self::{dolor::foo, ipsum::bar};
+
+use super::{
+    schema::{Context, Payload},
+    update::convert_publish_payload,
+};
+
+use crate::models::Event;
+
+use {
+    alloc::alloc::Layout,
+    broker::database::PooledConnection,
+    chrono::Utc,
+    core::f32,
+    juniper::{FieldError, FieldResult},
+    std::sync::Arc,
+    uuid::Uuid,
+};
diff --git a/tests/target/configs/group_imports/ByDistanceDescending.rs b/tests/target/configs/group_imports/ByDistanceDescending.rs
new file mode 100644
index 00000000..f3d4e3d1
--- /dev/null
+++ b/tests/target/configs/group_imports/ByDistanceDescending.rs
@@ -0,0 +1,23 @@
+// rustfmt-group_imports: ByDistanceDescending
+
+mod dolor;
+mod ipsum;
+
+use {
+    alloc::alloc::Layout,
+    broker::database::PooledConnection,
+    chrono::Utc,
+    core::f32,
+    juniper::{FieldError, FieldResult},
+    std::sync::Arc,
+    uuid::Uuid,
+};
+
+use crate::models::Event;
+
+use super::{
+    schema::{Context, Payload},
+    update::convert_publish_payload,
+};
+
+use self::{dolor::foo, ipsum::bar};
diff --git a/tests/target/configs/group_imports/dolor/mod.rs b/tests/target/configs/group_imports/dolor/mod.rs
new file mode 100644
index 00000000..8b137891
--- /dev/null
+++ b/tests/target/configs/group_imports/dolor/mod.rs
@@ -0,0 +1 @@
+
diff --git a/tests/target/configs/group_imports/ipsum/mod.rs b/tests/target/configs/group_imports/ipsum/mod.rs
new file mode 100644
index 00000000..8b137891
--- /dev/null
+++ b/tests/target/configs/group_imports/ipsum/mod.rs
@@ -0,0 +1 @@
+
diff --git a/tests/target/configs/regroup_modules/dolor/mod.rs b/tests/target/configs/regroup_modules/dolor/mod.rs
new file mode 100644
index 00000000..8b137891
--- /dev/null
+++ b/tests/target/configs/regroup_modules/dolor/mod.rs
@@ -0,0 +1 @@
+
diff --git a/tests/target/configs/regroup_modules/false.rs b/tests/target/configs/regroup_modules/false.rs
new file mode 100644
index 00000000..72e36a2d
--- /dev/null
+++ b/tests/target/configs/regroup_modules/false.rs
@@ -0,0 +1,8 @@
+// rustfmt-regroup_modules: false
+// Regroup modules
+
+mod dolor;
+mod lorem;
+
+pub mod ipsum;
+mod sit;
diff --git a/tests/target/configs/regroup_modules/ipsum/mod.rs b/tests/target/configs/regroup_modules/ipsum/mod.rs
new file mode 100644
index 00000000..8b137891
--- /dev/null
+++ b/tests/target/configs/regroup_modules/ipsum/mod.rs
@@ -0,0 +1 @@
+
diff --git a/tests/target/configs/regroup_modules/lorem/mod.rs b/tests/target/configs/regroup_modules/lorem/mod.rs
new file mode 100644
index 00000000..8b137891
--- /dev/null
+++ b/tests/target/configs/regroup_modules/lorem/mod.rs
@@ -0,0 +1 @@
+
diff --git a/tests/target/configs/regroup_modules/sit/mod.rs b/tests/target/configs/regroup_modules/sit/mod.rs
new file mode 100644
index 00000000..8b137891
--- /dev/null
+++ b/tests/target/configs/regroup_modules/sit/mod.rs
@@ -0,0 +1 @@
+
diff --git a/tests/target/configs/regroup_modules/true.rs b/tests/target/configs/regroup_modules/true.rs
new file mode 100644
index 00000000..064d7fbb
--- /dev/null
+++ b/tests/target/configs/regroup_modules/true.rs
@@ -0,0 +1,7 @@
+// rustfmt-regroup_modules: true
+// Regroup modules
+
+mod dolor;
+pub mod ipsum;
+mod lorem;
+mod sit;
